File: algorithms_toolbox/week3_greedy_algorithms/3_car_fueling/CarFueling.java
import java.util.*;

public class CarFueling {
    static int computeMinRefills(int dist, int tank, int[] stops) {
    	int count = 0; 
    	int currRefill = 0;
    	int n = stops.length-2;
    	while(currRefill <= n) {
    		int lastRefill = currRefill;
    		while((currRefill <= n) && (stops[currRefill+1] - stops[lastRefill] <= tank)) {
    			currRefill++;
    		}
    	if (currRefill == lastRefill) {return -1;}
    	if (currRefill <= n) {
    		count++;
    		}
    	}
    	return count;
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int dist = scanner.nextInt();
        int tank = scanner.nextInt();
        int n = scanner.nextInt();
        int stops[] = new int[n+2];
        for (int i = 1; i <= n; i++) {
            stops[i] = scanner.nextInt();
        }
        stops[n+1] = dist;
        System.out.println(computeMinRefills(dist, tank, stops));
        scanner.close();
    }
}

File: algorithms_toolbox/week3_greedy_algorithms/3_car_fueling/car_fueling.py
# python3
import sys
from collections import deque

def compute_min_refills(distance, tank, stops):
    # write your code here
    stops.append(distance)
    
    count = 0
    prev = 0
    if stops[0]>tank:
        return -1
    tmp_dist = 0
    for i in range(0, len(stops)-1):
        if stops[i+1]-stops[i] > tank:
            return -1
        if stops[i+1]-tmp_dist <= tank:
            continue
        prev = i
        tmp_dist=stops[prev]
        count += 1
        #print("stops", stops, "prev", prev, "count", count)
    return count

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    d = data[0]
    m = data[1]
    n = data[2]
    stops = data[3:]
    print(compute_min_refills(d, m, stops))
